# NYC Yellow Taxi Trip Data Cleaning and Exploratory Analysis (Python)

In [61]:
# Import necessary libraries
import pandas as pd
import numpy as np

In [62]:
# Load the dataset and display the first 10 rows
df = pd.read_csv('taxi_data.csv')
df.head(10)   # By default, displays first 5 rows

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5    23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6    37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7    69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8     8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9    95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   
5                6           2.30           1                  N   
6                1          12.83           1                  N   
7                1           2.98           1                  N   
8                1           1.20           1                  N   
9                1           1.60           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   
5           161           236             1          9.0    0.5      0.5   
6            79           241             1         47.5    1.0      0.5   
7           237           114             1         16.0    1.0      0.5   
8           234           249             2          9.0    0.0      0.5   
9           239           237             1         13.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80  
5        2.06           0.0                    0.3         12.36  
6        9.86           0.0                    0.3         59.16  
7        1.78           0.0                    0.3         19.58  
8        0.00           0.0                    0.3          9.80  
9        2.75           0.0                    0.3         16.55

In [63]:
# Concise summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [64]:
# summary statistics of the numerical columns
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

In [65]:
# Check for missing values in each column
df.isnull().sum()

Unnamed: 0               0
VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
dtype: int64

In [66]:
# Drop Unnamed column as it is not needed
df.drop(columns=['Unnamed: 0'], inplace=True)

# Check the first 5 rows again to confirm the column has been dropped
df.head()

VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  passenger_count  \
0         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM                6   
1         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM                1   
2         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM                1   
3         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM                1   
4         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM                1   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.34           1                  N           100           231   
1           1.80           1                  N           186            43   
2           1.00           1                  N           262           236   
3           3.70           1                  N           188            97   
4           4.37           1                  N             4           112   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         13.0    0.0      0.5        2.76           0.0   
1             1         16.0    0.0      0.5        4.00           0.0   
2             1          6.5    0.0      0.5        1.45           0.0   
3             1         20.5    0.0      0.5        6.39           0.0   
4             2         16.5    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  
0                    0.3         16.56  
1                    0.3         20.80  
2                    0.3          8.75  
3                    0.3         27.69  
4                    0.3         17.80

In [67]:
# Convert two 'object' type variables 'tpep_pickup_datetime', and 'tpep_dropoff_datetime' to datetime format
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime'])

# Verify the conversion by checking the data types again
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[ns]
tpep_dropoff_datetime    datetime64[ns]
passenger_count                   int64
trip_distance                   float64
RatecodeID                        int64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
dtype: object

In [69]:
# Sort the variable 'trip_distance' in descending order and display the top 5 rows
# df.sort_values(by='trip_distance', ascending=False).head()
df_sorted = df.sort_values(by='trip_distance', ascending=False)
df_sorted.head(10)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
9280          2  2017-06-18 23:33:25   2017-06-19 00:12:38                2   
13861         2  2017-05-19 08:20:21   2017-05-19 09:20:30                1   
6064          2  2017-06-13 12:30:22   2017-06-13 13:37:51                1   
10291         2  2017-09-11 11:41:04   2017-09-11 12:18:58                1   
29            2  2017-11-06 20:30:50   2017-11-07 00:00:00                1   
18130         1  2017-10-26 14:45:01   2017-10-26 16:12:49                1   
5792          2  2017-08-11 14:14:01   2017-08-11 15:17:31                1   
15350         2  2017-09-14 13:44:44   2017-09-14 14:34:29                1   
10302         1  2017-05-15 08:11:34   2017-05-15 09:03:16                1   
2592          2  2017-06-16 18:51:20   2017-06-16 19:41:42                1   

       trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
9280           33.96           5                  N           132   
13861          33.92           5                  N           229   
6064           32.72           3                  N           138   
10291          31.95           4                  N           138   
29             30.83           1                  N           132   
18130          30.50           1                  N           132   
5792           30.33           2                  N           132   
15350          28.23           2                  N            13   
10302          28.20           2                  N            90   
2592           27.97           2                  N           261   

       DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
9280            265             2       150.00    0.0      0.0        0.00   
13861           265             1       200.01    0.0      0.5       51.64   
6064              1             1       107.00    0.0      0.0       55.50   
10291           265             2       131.00    0.0      0.5        0.00   
29               23             1        80.00    0.5      0.5       18.56   
18130           220             1        90.50    0.0      0.5       19.85   
5792            158             1        52.00    0.0      0.5       14.64   
15350           132             1        52.00    0.0      0.5        4.40   
10302           132             1        52.00    0.0      0.5       11.71   
2592            132             2        52.00    4.5      0.5        0.00   

       tolls_amount  improvement_surcharge  total_amount  
9280           0.00                    0.3        150.30  
13861          5.76                    0.3        258.21  
6064          16.26                    0.3        179.06  
10291          0.00                    0.3        131.80  
29            11.52                    0.3        111.38  
18130          8.16                    0.3        119.31  
5792           5.76                    0.3         73.20  
15350          5.76                    0.3         62.96  
10302          5.76                    0.3         70.27  
2592           5.76                    0.3         63.06

In [94]:
# Sort the second variable 'total_amount' in descending order and display the top 5 rows
# df.sort_values(by='total_amount', ascending=False).head()
df_sorted_amount = df.sort_values(['total_amount'], ascending=False)['total_amount']
df_sorted_amount.head(20)

8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
1928      137.80
10291     131.80
6708      126.00
11608     123.30
908       121.56
7281      120.96
18130     119.31
13621     115.94
13359     111.95
29        111.38
Name: total_amount, dtype: float64

In [75]:
# Check the last 20 values of the 'total_amount' column to see if there are any negative or zero values
df_sorted_amount.tail(20)

14283      0.31
19067      0.30
10506      0.00
5722       0.00
4402       0.00
22566      0.00
1646      -3.30
18565     -3.80
314       -3.80
5758      -3.80
5448      -4.30
4423      -4.30
10281     -4.30
8204      -4.80
20317     -4.80
11204     -5.30
14714     -5.30
17602     -5.80
20698     -5.80
12944   -120.30
Name: total_amount, dtype: float64

In [76]:
# How many of each payment type are there?
df['payment_type'].value_counts()

payment_type
1    15265
2     7267
3      121
4       46
Name: count, dtype: int64

In [88]:
# What is the average tip for trips paid by 'credit card'
average_cc_tip = df[df['payment_type'] == 1]['tip_amount'].mean()
print('Average Credit Card Tip:' , average_cc_tip)

# What is the average tip for trips paid by 'cash'
average_cash_tip = df[df['payment_type']==2]['tip_amount'].mean()
print('Average Cash Tip:' , average_cash_tip)

Average Credit Card Tip: 2.7298001965280054
Average Cash Tip: 0.0


In [95]:
# How many time of each VendorID represented in the data
df['VendorID'].value_counts()

VendorID
2    12626
1    10073
Name: count, dtype: int64

In [91]:
# Mean value of 'total_amount' for each vendor
df.groupby('VendorID').mean(numeric_only=True)[['total_amount']]

total_amount
VendorID              
1            16.298119
2            16.320382

In [92]:
# Filter the data for credit card payments only
credit_card = df[df['payment_type'] == 1]

# Filter the only credit card data with passengers count
credit_card['passenger_count'].value_counts()

passenger_count
1    10977
2     2168
5      775
3      600
6      451
4      267
0       27
Name: count, dtype: int64

In [96]:
# Calculate the average tip amount for each passenger count in credit card payments
credit_card.groupby(['passenger_count']).mean(numeric_only=True)[['tip_amount']]

tip_amount
passenger_count            
0                  2.610370
1                  2.714681
2                  2.829949
3                  2.726800
4                  2.607753
5                  2.762645
6                  2.643326

# Summary

1. Import necessary libraries
2. Load the dataset and display the first 10 rows
3. Get a concise summary of the DataFrame including the data types and non-null values. Ther are 18 columns in total and 22699 rows. The data types include integers, floats, and objects (strings). 
4. Show summary statistics of the numerical columns in the DataFrame, i.e. count, mean, std, min, 25%, 50%, 75%, and max values.
5. Check for missing values in each column to identify any potential data quality issues. There are no missing values in any of the columns.
6. delete column 'Unnamed: 0' as it is not needed.
7. Convert two 'object' type variables 'tpep_pickup_datetime', and 'tpep_dropoff_datetime' to datetime format for easier manipulation and analysis.
8. Sort the variable 'trip_distance' in descending order and display the top 10 rows to identify the longest trips in the dataset. We noticed that some trips have extremely high distances, which may indicate potential outliers or data entry errors.
9. Sort the second variable 'total_amount' in descending order and display the top 15 rows to identify the highest fare amounts in the dataset. Similar to trip distances, some fare amounts are unusually high, suggesting possible outliers or data inaccuracies.
10. Check the last 20 values of the 'total_amount' column to see if there are any negative or zero values, which would be invalid for fare amounts. We found several entries with negative or zero total amounts, indicating potential data quality issues that need to be addressed.
11. Check the distribution of payment types used in the dataset by counting the occurrences of each payment type. This helps to understand the preferred payment methods among passengers. The most common payment type is '1', followed by '2', '3', and '4' which is 'credit card' payment.
12. Tip Analysis Based on Payment Type:
    - Calculate the average tip amount for trips paid by 'credit card' (payment type 1).
    - Calculate the average tip amount for trips paid by 'cash' (payment type 2).
    - Compare the average tips for both payment types to see if there is a significant difference in tipping behavior based on the payment method. We found that the average tip for credit card payments is higher than that for cash payments, suggesting that passengers may be more inclined to tip generously when using credit cards.
13. Vendor Analysis: Count the number of trips serviced by each vendor to understand their market share in the dataset. This analysis helps to identify which vendors are more active in providing taxi services. We found that VendorID '2' has serviced more trips compared to VendorID '1', indicating a larger market presence for VendorID '2'.
14. Filter the data for credit card payments only and analyze the distribution of passenger counts for these trips. 